htlcswitch/link.go

```diff
@@ -1675,6 +1675,28 @@ func (l *channelLink) handleUpstreamMsg(msg lnwire.Message) {
 	case *lnwire.UpdateFulfillHTLC:
 		pre := msg.PaymentPreimage
 		idx := msg.ID
+
+		// Before we pipeline the settle, we'll check the set of active
+		// htlc's to see if the related UpdateAddHTLC has been fully
+		// locked-in.
+		var lockedin bool
+		htlcs := l.channel.ActiveHtlcs()
+		for _, add := range htlcs {
+			// The HTLC will be outgoing and match idx.
+			if !add.Incoming && add.HtlcIndex == idx {
+				lockedin = true
+				break
+			}
+		}
+
+		if !lockedin {
+			l.fail(
+				LinkFailureError{code: ErrInvalidUpdate},
```

Comment:
Bolt 2 language regarding `update_fulfill_htlc`: https://github.com/lightning/bolts/blob/master/02-peer-protocol.md#requirements-10

- bullet point 3 says that the sending node MUST NOT send update_fulfill_htlc until the add is locked-in. The language of MUST NOT is confusing, does this mean the recipient has to force close? IMO following the spec here to the letter may be a little superfluous as the sender shouldn't be sending this in the first place, and the receiver did nothing wrong. So I believe it's fine to not force close and instead shut down the link.
- Under the receiving node requirements _"If the id does not correspond to an HTLC in its current commitment transaction:...MUST send a warning and close the connection, or send an error and fail the channel."_
  - Again, the sender should not be doing this. We don't close the channel or send a warning, but we do shut down the link and send an error. From lnd's PoV errors are benign, but I believe that other implementations force close upon receiving the error.
  - If other implementations would force close _anyways_ due to our error, does it make sense to just make this `l.fail` a force close?

Reply:
> So I believe it's fine to not force close and instead shut down the link.

Given that if a commitment sig _isn't_ sent after/during this early fulfill, IMO just taking down the link lets things resume vs the hard route of needing to eat those chain fees.